sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -410,10 +428,31 @@ Scheduler::~Scheduler() {
           "not all resources were released. Please be sure that all kernels "
           "have synchronization points.\n\n");
   }
+  // Please be aware that releaseResources should be called before deletion of
+  // Scheduler. Otherwise there can be the case when objects Scheduler keeps as
+  // fields may need Scheduler for their release and they work with Scheduler
+  // via GlobalHandler::getScheduler that will create new Scheduler object.
+  // Still keep it here but it should do almost nothing if releaseResources
+  // called before.
+  releaseResources();
+}
+
+void Scheduler::releaseResources() {
   // There might be some commands scheduled for post enqueue cleanup that
   // haven't been freed because of the graph mutex being locked at the time,
   // clean them up now.
   cleanupCommands({});
+  DefaultHostQueue.reset();
+
+  // We need loop since sometimes we may need new objects to be added to
+  // deferred mem objects storage during cleanup. Known example is: we cleanup
+  // existing deferred mem objects under write lock, during this process we
+  // cleanup commands related to this record, command may have last reference to
+  // queue_impl, ~queue_impl is called and buffer for assert (which is created
+  // with size only so all confitions for deferred release are satisfied) is
+  // added to deferred mem obj storage. So we may end up with leak.
+  while (!isDeferredMemObjectsEmpty())
+    cleanupDeferredMemObjects(true);
```

Comment:
```suggestion
    cleanupDeferredMemObjects(/*Blocking*/true);
```
Minor.

Reply:
fixed in https://github.com/intel/llvm/pull/6837/commits/d4537a3c85672ebd6fe7ccee110b98c5d3f46b5b